### PR TITLE
Change formatting of timestamp to be closer to ISO8601

### LINKED ISSUE #4560: log_time format is different between documentation and respone
## Problem Definition

I am calling the payment REST API:
`http://localhost:5001/api/v1/payments/0x95B2d84De40a0121061b105E6B54016a49621B44/0x82641569b2062B545431cF6D7F0A418582865ba7`

According to the documentation:
```
raiden_events contain a timestamp field, log_time, indicating when they were written to the write-ahead log. The format of log_time is ISO8601 with milliseconds.
```
Verifying with v0.100.1 this seems to be true for that version but it is not for the latest develop version.

What the API returns now is the following: `"log_time": "2019-08-06 14:52:02"`. This format is also what is stored in the database for the timestamps.

```json
[
    {
        "event": "EventPaymentReceivedSuccess",
        "log_time": "2019-08-06 14:52:02",
        "identifier": 1565103119921,
        "amount": 1,
        "initiator": "0x82641569b2062B545431cF6D7F0A418582865ba7"
    }
]
```

### System Description

Here add a detailed description of your system, e.g. output of the following script:

```
Linux forseti 5.2.6-arch1-1-ARCH #1 SMP PREEMPT Sun Aug 4 14:58:49 UTC 2019 x86_64 GNU/Linux
/usr/bin/solc
solc, the solidity compiler commandline interface
Version: 0.5.10+commit.5a6ea5b1.Linux.g++
/usr/bin/geth
INFO [08-09|13:31:41.596] Bumping default cache on mainnet         provided=1024 updated=4096
Geth
Version: 1.9.1-stable
Git Commit: b7b2f60f866d92e75db5bfe5be51b48eadeb3150
Git Commit Date: 20190724
Architecture: amd64
Protocol Versions: [63]
Network Id: 1
Go Version: go1.12.7
Operating System: linux
GOPATH=
GOROOT=/usr/lib/go
/usr/bin/parity
Parity Ethereum
  version Parity-Ethereum/v2.5.5-stable-3ebc769757-20190708/x86_64-linux-gnu/rustc1.35.0
Copyright 2015-2019 Parity Technologies (UK) Ltd.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

By Wood/Paronyan/Kotewicz/Drwięga/Volf
   Habermeier/Czaban/Greeff/Gotchac/Redmann

/home/kelsos/.virtualenvs/raiden-develop/bin/raiden
{
  "raiden": "0.100.5.dev288+ga2270259b",
  "raiden_db_version": 23,
  "python_implementation": "CPython",
  "python_version": "3.7.4",
  "system": "Linux 64bit_ELF 5.2.6-arch1-1-ARCH",
  "architecture": "x86_64",
  "distribution": "source"
}

```

### PR DESCRIPTION
**Pull request by lullis**
The API documentation states that records that provide a timestamp should follow the ISO8601 spec.
Meanwhile, our database tables were defined as storing the default `CURRENT_TIMESTAMP` format from sqlite, which do not hold milisecond information.

This commit changes all tables to start using TIMESTAMP fields with formatted with milisecond as well.

Fixes #4560

Quality check list that cannot be automatically verified.

- Safety
    - [ ] The changes respect the necessary conditions for safety (https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints)
-  Code quality
    - [ ] Error conditions are handled
    - [ ] Exceptions are propagated to the correct parent greenlet
    - [ ] Exceptions are correctly classified as recoverable or unrecoverable
- Compatibility
    - [ ] State changes are forward compatible
    - [ ] Transport messages are backwards and forward compatible
- Commits
    - [ ] Have good messages
    - [ ] Squashed unecessary commits
- If it's a bug fix:
    - Regression test for the bug
        - [ ] Properly covers the bug
        - [ ] If an integration test is used, it could not be written as a unit test
- Is it a user facing feature/bug fix?
    - [ ] Changelog entry has been added

### CODE REVIEW

**Comment by LefterisJP** [comment]
Can you add a unit test for this? Should not be hard to do. Also this is a user facing bug and as such needs a changelog entry.

**Comment by lullis** [comment]
Honestly, If we just changing the default format of the timestamp field, I don't know what kind of unit test could be written that wouldn't just be testing sqlite itself.

However, one thing that I was considering when I looked into this code was to create a proper "TimestampField" for our marshmallow custom fields, which would allow us to more easily verify the data that is going through the API. I will look into this, but it will make this PR more than a simple change.

**Comment by LefterisJP** [comment]
> I don't know what kind of unit test could be written that wouldn't just be testing sqlite itself.

Hmmm? What do you mean? The bug was found since the expected timestamp format in the docs did not match the timestamp saved in raiden and the DB.

Just write a super small test that makes sure the documented timestamp format is what we are using.

**Comment by lullis** [comment]
I am failing to find any other example of unit test where we are checking for field formatting, and the tests in the API are also concerning with test schemas/marshmallow things. I will look into it.

File: raiden/storage/utils.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def __getattr__(self, item):
 CREATE TABLE IF NOT EXISTS state_changes (
     identifier ULID PRIMARY KEY NOT NULL,
     data JSON,
-    timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
+    timestamp TIMESTAMP DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')) NOT NULL
```
Comment (hackaugusto) [inline raiden/storage/utils.py]: This is not valid ISO8601

Comment (hackaugusto) [reply raiden/storage/utils.py]: > In formats 5 through 7, the "T" is a literal character separating the date and the time, as required by ISO-8601

https://www.sqlite.org/lang_datefunc.html

Comment (LefterisJP) [reply raiden/storage/utils.py]: +1 you are missing the seconds there

Comment (LefterisJP) [reply raiden/storage/utils.py]: Just make sure that all the examples in the docs and what you implement here follow ISO8601:  https://raiden-network.readthedocs.io/en/latest/rest_api.html#get--api-v1-payments-(token_address)-(target_address)

Comment (lullis) [reply raiden/storage/utils.py]: Yeap, missing the T to separate date from time. There was some issue with the sqlite library itself when it was trying to actually parse the string with the `T` in there. I will try to repro it and paste it here.

**Comment by hackaugusto** [comment]
> I am failing to find any other example of unit test where we are checking for field formatting, and the tests in the API are also concerning with test schemas/marshmallow things. I will look into it.

Just write a state change / event and then load it. It will not just test the sqlite implementation but also our data model. It was working before but because of the lack of the tests my changes broke it.

It would also be nice to have it in the REST API tests, because it can go wrong on other places.

File: raiden/storage/utils.py (hunk 1 above)
Comment (lullis) [reply raiden/storage/utils.py]: @hackaugusto: got to repro the error when placing `T` in the strftime string. apparently sqlite does not follow ISO here and expects a space to separate date from time.

```$ pytest raiden/tests/unit/ --lf
============================================================================================================================= test session starts =============================================================================================================================
platform linux -- Python 3.7.3, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/raphael/code/brainbot/raiden/client/.hypothesis/examples')
rootdir: /home/raphael/code/brainbot/raiden/client, inifile: setup.cfg
plugins: xdist-1.28.0, forked-1.0.2, random-0.2, select-0.1.2, hypothesis-4.24.3
collected 25 items / 23 deselected / 2 selected                                                                                                                                                                                                                               
run-last-failure: rerun previous 2 failures (skipped 1134 files)

raiden/tests/unit/test_sqlite.py F                                                                                                                                                                                                                                      [ 50%]
raiden/tests/unit/test_wal.py F                                                                                                                                                                                                                                         [100%]

================================================================================================================================== FAILURES ===================================================================================================================================
________________________________________________________________________________________________________________________________ test_log_run _________________________________________________________________________________________________________________________________

    def test_log_run():
        with patch("raiden.storage.sqlite.get_system_spec") as get_speck_mock:
            get_speck_mock.return_value = dict(raiden="1.2.3")
            serializer = JSONSerializer()
            store = SerializedSQLiteStorage(":memory:", serializer)
            store.log_run()
        cursor = store.database.conn.cursor()
>       cursor.execute("SELECT started_at, raiden_version FROM runs")

raiden/tests/unit/test_sqlite.py:372: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

val = b'2019-08-14T10:38:26.447'

    def convert_timestamp(val):
>       datepart, timepart = val.split(b" ")
E       ValueError: not enough values to unpack (expected 2, got 1)

/usr/lib/python3.7/sqlite3/dbapi2.py:67: ValueError
___________________________________________________________________________________________________________________________ test_write_read_events ____________________________________________________________________________________________________________________________

    def test_write_read_events():
        wal = new_wal(state_transition_noop)
    
        event = EventPaymentSentFailed(
            factories.make_token_network_registry_address(),
            factories.make_address(),
            1,
            factories.make_address(),
            "whatever",
        )
    
        with pytest.raises(sqlite3.IntegrityError):
            unexisting_state_change_id = random.getrandbits(16 * 8).to_bytes(16, "big")
            wal.storage.write_events([(unexisting_state_change_id, event)])
    
        previous_events = wal.storage.get_events_with_timestamps()
    
        state_change_ids = wal.storage.write_state_changes([StateChange()])
        wal.storage.write_events([(state_change_ids[0], event)])
    
>       new_events = wal.storage.get_events_with_timestamps()

raiden/tests/unit/test_wal.py:164: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
raiden/storage/sqlite.py:859: in get_events_with_timestamps
    limit=limit, offset=offset, filters=filters, logical_and=logical_and
raiden/storage/sqlite.py:662: in get_events_with_timestamps
    limit=limit, offset=offset, filters=filters, logical_and=logical_and
raiden/storage/sqlite.py:594: in _query_events
    logical_and=logical_and,
raiden/storage/sqlite.py:475: in _form_and_execute_json_query
    cursor.execute(query, args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

val = b'2019-08-14T10:38:26.526'

    def convert_timestamp(val):
>       datepart, timepart = val.split(b" ")
E       ValueError: not enough values to unpack (expected 2, got 1)

/usr/lib/python3.7/sqlite3/dbapi2.py:67: ValueError
=================================================================================================================== 2 failed, 23 deselected in 0.55 seconds```

@LefterisJP, not really. sqlite strftime is different from python. `%f` represents *seconds and miliseconds*

Comment (LefterisJP) [reply raiden/storage/utils.py]: > @LefterisJP, not really. sqlite strftime is different from python. `%f` represents *seconds and miliseconds*

Interesting. I did not know that. Which is why having tests to make sure all of this works as expected is needed.

Comment (lullis) [reply raiden/storage/utils.py]: I added an integration test now. The "T" is still missing on the timestamp string to actually be considered a proper ISO8601 string. However I tested with both Python and Javascript and they both managed to parse the string with a blank space.

To actually get a fully-compliant string, the only sane-ish approach that I can think now is to to turn the timestamp on sqlite to regular string fields and some marshmallow massaging of the fields.

If we are okay with having a working-but-not-really-ISO timestamp, then I will also write an unit test and it should be okay to merge.

Comment (palango) [reply raiden/storage/utils.py]: Can you add a unit test as well? Seems like there should be a quick way of testing this without starting the whole integration test stuff

**Comment by lullis** [comment]
@LefterisJP, @hackaugusto. I took the approach of adding a timestamp-specific class of marshmallow field, so that the datetime representation of sqlite does not affect our desired representation of the string on the REST API. Please take a look at the updates and let me know if there is any more tests that are needed.

File: raiden/api/v1/encoding.py [hunk 2]
```diff
@@ -78,6 +80,20 @@ def _deserialize(self, value, attr, data, **kwargs):  # pylint: disable=unused-a
         return value
 
 
+class TimeStampField(fields.DateTime):
```
Comment (palango) [inline raiden/api/v1/encoding.py]: Why can't you use `DateTime` directly? AFAICT it does the same thing.

Comment (lullis) [reply raiden/api/v1/encoding.py]: The issue is with sqlite DB API and the paste with errors from the test runner above show what happens. The timestamp fields are represented as strings on sqlite, but python's dbapi will try to turn them into datetime objects implicitly. Given that sqlite does not place the "T" as the separator between the date and time sections of the timestamp, the parsing fails. See the error from the test a few comments above.

This custom field bypasses the dbapi implicit conversion by treating the sqlite field as a simple string when serializing/deserializing, and doing the parsing ourselves.

Comment (palango) [reply raiden/api/v1/encoding.py]: That's not my point. For me it seems that we don't need a custom field. If I look at https://marshmallow.readthedocs.io/en/3.0/_modules/marshmallow/fields.html#DateTime I see that it handles `None` values as your implementation does, and also uses `iso` as the default format.

Comment (lullis) [reply raiden/api/v1/encoding.py]: I am trying to explain what happened when I *did use DateTime directly*.

 1. The deserialization part failed and I all got was some stacktrace showing that the part of the code that was throwing the exception was at sqlite dbapi.
 1. The serialilzation part did not throw any errors, but the rendered timestamp had the timezone modifier.

GIven these points, I don't see how we could use the DateTime field directly.

**Pull request merged**